refactor: move from io/ioutil to io and os packages

Juneezee opened the pull request:
## Description of the Pull Request (PR):

The `io/ioutil` package has been deprecated in Go 1.16 (See https://pkg.go.dev/io/ioutil). This PR replaces the existing `io/ioutil` functions with their new definitions in `io` and `os` packages.

- `ioutil.Discard` => `io.Discard`
- `ioutil.NopCloser` => `io.NopCloser`
- `ioutil.ReadAll` => `io.ReadAll`
- `ioutil.ReadDir` => `os.ReadDir` (returns a slice of `os.DirEntry` rather than a slice of `fs.FileInfo`, this may improve performance in some cases as `os.ReadDir` is more efficient [\[1\]])
- `ioutil.ReadFile` => `os.ReadFile`
- `ioutil.TempDir` => `os.MkdirTemp`
- `ioutil.TempFile` => `os.CreateTemp`
- `ioutil.WriteFile` => `os.WriteFile`

[\[1\]]: https://pkg.go.dev/io/ioutil#ReadDir


### This fixes or addresses the following GitHub issues:


#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

DrDaveD commented:
Thanks for the contribution, @Juneezee.  I noticed that SingularityCE has a similar pending PR too at sylabs/singularity#964.

DrDaveD commented:
Sorry, meanwhile another merge has caused a conflict.  Please rebase on the current main.

Juneezee commented:
> Sorry, meanwhile another merge has caused a conflict. Please rebase on the current main.

@DrDaveD No worries. I've rebased onto `main`. Please take a look again. Thanks :smiley:

DrDaveD commented:
@cclerget agreed to review this when he has a chance

cclerget reviewed (changes requested):
Thank you for tackling this technical debt. I have one change request about the RPC `ReadDir` method which must return `os.FileInfo` otherwise LGTM. Thanks !

Juneezee commented:
@cclerget Hi, thanks for reviewing this PR. I've reverted RPC `ReadDir` to return `os.FileInfo` and also rebased onto `main`. Please take a look again. Thanks

cclerget commented on internal/pkg/image/driver/imagedriver.go:
```diff
@@ -346,7 +345,8 @@ func (d *fuseappsDriver) Start(params *image.DriverParams, containerPid int) err
 
 // Stop the process associated with the mount target, if there is one.
 // If kill is not true, an unmount should already have happened so at
-//   first just wait for the process to exit.
```
I didn't spotted this at first, can you also remove the space from the comment to avoid having go 1.19 formatting it to:

```
//
//	first just wait for the process to exit.
```

Thanks !

cclerget replied:
@Juneezee ^^ , remain just this one and it's good to go

cclerget reviewed (approved):
Great thank you very much @Juneezee !
